### PR TITLE
use logsigmoid at multilabel_soft_margin_loss, and change output from shape=(N, C)to (N,)

### LINKED ISSUE #9141: Make torch.nn.functional.multilabel_soft_margin_loss more stable
Currently it calls `sigmoid` and then `binary_cross_entropy` (which does `log` internally):
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L1796-L1797

Is it not possible to compute it as: `torch.mean(-(y * F.logsigmoid(x) + (1 - y) * F.logsigmoid(-x)))` ?

```
y * log(1 / (1 + exp(-x)) + (1 - y) * log(1 - 1 / (1 + exp(-x))) =
y * logsigmoid(x) + (1 - y) * log((1 + exp(-x) - 1) / (1 + exp(-x))) = 
y * logsigmoid(x) + (1 - y) * log(1 / (1 + exp(x))) = 
y * logsigmoid(x) + (1 - y) * logsigmoid(-x)
```

### PR DESCRIPTION
**Pull request by weiyangfb**
- fixes #9141, #9301
- use logsigmoid at multilabel_soft_margin_loss to make it more stable (NOT fixing legacy MultiLabelSoftMarginCriterion)
- return (N) instead of (N, C) to match the same behavior as MultiMarginLoss
- Note that with this PR, the following behavior is expected:
```
loss = F.multilabel_soft_margin_loss(outputs, labels, reduction='none')
loss_mean = F.multilabel_soft_margin_loss(outputs, labels, reduction='elementwise_mean')
loss_sum = F.multilabel_soft_margin_loss(outputs, labels, reduction='sum')

loss.sum() == loss_sum  # True
loss.mean() == loss_mean  # True
```

### CODE REVIEW

File: torch/nn/functional.py [hunk 1]
```diff
@@ -1777,8 +1777,24 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
...
+    loss = - (target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
+
+    if weight is not None:
+        loss = loss * weight
+
+    if reduction != 'none':
+        if reduction == 'sum':
+            return loss.sum()
+        else:  # elementwise_mean
+            return loss.mean()
```
Comment (soumith) [inline torch/nn/functional.py]: this is not element-wise mean, it's global mean.

**Review by soumith — APPROVED** [review]
approved. change comment.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -1777,8 +1777,24 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
     """
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    input = torch.sigmoid(input)
-    return binary_cross_entropy(input, target, weight, None, None, reduction)
+
+    loss = - (target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
```
Comment (soumith) [inline torch/nn/functional.py]: Isn't it
```
loss = - ((target * logsigmoid(input) + (1 - target) * logsigmoid(-input)))
```

i.e. negative sign comes outer?

File: torch/nn/functional.py (hunk 1 above)
Comment (weiyangfb) [reply torch/nn/functional.py]: The [doc](https://github.com/pytorch/pytorch/blob/bfe2aa093e17c84db863672ad446de064e47dc7e/torch/nn/modules/loss.py#L889-L894) and [test](https://github.com/pytorch/pytorch/blob/9a9a7325c6f6b7f82636b6efdd742c9d1e63584e/test/test_nn.py#L6166-L6176) suggest this should be a global mean. So yeah, the args name is confusing...

File: torch/nn/functional.py [hunk 3]
```diff
@@ -1777,8 +1777,24 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
...
+
+    if weight is not None:
+        loss = loss * weight
+
+    if reduction != 'none':
```
Comment (li-roy) [inline torch/nn/functional.py]: The two lines at the top of the function parses reduce and size_average into reduction, so you only need to handle the cases for reduction.

File: torch/nn/functional.py [hunk 4]
```diff
@@ -1777,8 +1777,24 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
...
+        if reduction == 'sum':
+            return loss.sum()
+        else:  # global mean
+            return loss.mean()
```
Comment (li-roy) [inline torch/nn/functional.py]: Let's resolve #9301 and #10034 (both are the same issue) by changing the behavior to match the docs (divide by `size(0)`).

**Comment by li-roy** [comment]
Why are we changing the behavior of all these losses?

File: torch/nn/functional.py [hunk 5]
```diff
@@ -1827,8 +1827,21 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
     """
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    input = torch.sigmoid(input)
-    return binary_cross_entropy(input, target, weight, None, None, reduction)
+
+    loss = -(target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
+    loss.sum(dim=0)  # only return N loss values
```
Comment (li-roy) [inline torch/nn/functional.py]: I think this should be something like
```loss = loss.sum(dim=1)```

**Review by li-roy — CHANGES_REQUESTED** [review]

File: torch/nn/functional.py (hunk 5 above)
Comment (weiyangfb) [reply torch/nn/functional.py]: oh, you are right, I will change that

**Comment by weiyangfb** [comment]
@li-roy I made changes accordingly, maybe take a look when you get a chance?

File: torch/nn/functional.py [hunk 6]
```diff
@@ -1827,8 +1827,22 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
...
+
+    loss = -(target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
+
+    if weight is not None:
+        loss = loss * weight
+
+    loss = loss.sum(dim=1)  # only return N loss values
+
+    if reduction == 'none':
+        return loss
```
Comment (li-roy) [inline torch/nn/functional.py]: I'm still a bit confused by the behavior we have here. It looks like if reduction is none, we're returning a tensor of N losses, and each individual loss is the sum of the losses for each class of that sample. However, when reduction is elementwise_mean or sum, it seems that you're now averaging the losses for each class, rather than summing them. We should choose one of these and stick to it, I believe summing is the legacy behavior.

**Review by li-roy — COMMENTED** [review]
Looks good for the most part, just need to be consistent in the behavior.

File: torch/nn/functional.py (hunk 6 above)
Comment (weiyangfb) [reply torch/nn/functional.py]: oh, you're right. Yeah, I will change that to follow the legacy behavior, so that 
```
loss = F.multilabel_soft_margin_loss(outputs, labels, reduction='none')
loss_mean = F.multilabel_soft_margin_loss(outputs, labels, reduction='elementwise_mean')
loss_sum = F.multilabel_soft_margin_loss(outputs, labels, reduction='sum')

loss.sum() == loss_sum  # True
loss.mean() == loss_mean  # True
```

Comment (li-roy) [reply torch/nn/functional.py]: Unless a loss is storing its elements in a weird way, the general rule of thumb is: if `loss` is the unreduced loss, then `loss.mean()` is the elementwise and `loss.sum()` is the sum.

**Comment by li-roy** [comment]
If you're planning to go with averaging across classes, don't forget to change the doc in torch/nn/modules/loss.py

**Review by li-roy — APPROVED** [review]

File: torch/nn/functional.py [hunk 7]
```diff
@@ -1827,8 +1827,22 @@ def multilabel_soft_margin_loss(input, target, weight=None, size_average=None,
     """
     if size_average is not None or reduce is not None:
         reduction = _Reduction.legacy_get_string(size_average, reduce)
-    input = torch.sigmoid(input)
-    return binary_cross_entropy(input, target, weight, None, None, reduction)
+
+    loss = -(target * logsigmoid(input) + (1 - target) * logsigmoid(-input))
```
Comment (vadimkantorov) [inline torch/nn/functional.py:1831]: this could probably use `torch.lerp` (yet AFAIK it fails export to ONNX and has NaN problems: https://github.com/pytorch/pytorch/issues/71701)
